
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Gnocis tutorial &#8212; Gnocis 0.9.0b documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="gnocis" href="gnocis.html" />
    <link rel="prev" title="License" href="license.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="gnocis.html" title="gnocis"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="license.html" title="License"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Gnocis 0.9.0b documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="gnocis-tutorial">
<h1>Gnocis tutorial<a class="headerlink" href="#gnocis-tutorial" title="Permalink to this headline">¶</a></h1>
<p>Bjørn Bredesen, 2018-2019</p>
<div class="section" id="set-up">
<h2>Set-up<a class="headerlink" href="#set-up" title="Permalink to this headline">¶</a></h2>
<p>Here, we will train a PREdictor (Ringrose <em>et al.</em> 2003) model on the Kahn <em>et al.</em> (2014) Polycomb targets, and predict candidate PREs genome-wide. We will go through all steps that must be executed, from loading a General Feature Format (GFF) file with genomic coordinates for Polycomb targets, extract the underlying sequences, train the model, calibrate the threshold, and predict candidate PREs genome-wide.</p>
<p>This tutorial assumes that Gnocis has been successfully installed.</p>
<p>We will use the <em>Drosophila melanogaster</em> genome throughout this tutorial. We have downloaded the version R5 of the genome from FlyBase (source URL: ftp://ftp.flybase.net/releases/FB2014_03/dmel_r5.57/fasta/dmel-all-chromosome-r5.57.fasta.gz), and included it with this tutorial for educational purposes. For copyright information regarding the <em>D. melanogaster</em> genome, see https://wiki.flybase.org/wiki/FlyBase:About#FlyBase_Copyright.</p>
</div>
<div class="section" id="preparing-training-data-using-gnocis">
<h2>Preparing training data using Gnocis<a class="headerlink" href="#preparing-training-data-using-gnocis" title="Permalink to this headline">¶</a></h2>
<p>Download the Gnocis repository, and in a terminal, navigate to the Gnocis <code class="docutils literal notranslate"><span class="pre">tutorial/</span></code> folder. Open the Python3 Read-Eval Print Loop (REPL) by running <code class="docutils literal notranslate"><span class="pre">python3</span></code>.</p>
<p>Run:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the library</span>
<span class="kn">import</span> <span class="nn">gnocis</span> <span class="kn">as</span> <span class="nn">cis</span>

<span class="c1"># We will use the Drosophila melanogaster genome for multiple examples. We define a stream of is to a variable. The stream ensures that only segments of the genome are loaded at a time, saving memory.</span>
<span class="n">Dmel</span> <span class="o">=</span> <span class="n">cis</span><span class="o">.</span><span class="n">streamFASTAGZ</span><span class="p">(</span><span class="s1">&#39;DmelR5.fasta.gz&#39;</span><span class="p">)</span>

<span class="c1"># Load the Kahn et al. (2014) Polycomb targets</span>
<span class="n">Kahn2014Rgn</span> <span class="o">=</span> <span class="n">cis</span><span class="o">.</span><span class="n">loadGFF</span><span class="p">(</span><span class="s1">&#39;Kahn2014.GFF&#39;</span><span class="p">)</span>

<span class="c1"># Resize and extract from genome</span>
<span class="n">Kahn2014Seq</span> <span class="o">=</span> <span class="n">Kahn2014Rgn</span>\
          <span class="o">.</span><span class="n">getRandomlyRecentered</span><span class="p">(</span><span class="mi">3000</span><span class="p">)</span>\
          <span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">Dmel</span><span class="p">)</span>

<span class="c1"># Split into independent training and test sets</span>
<span class="n">tpos</span><span class="p">,</span> <span class="n">vpos</span> <span class="o">=</span> <span class="n">Kahn2014Seq</span><span class="o">.</span><span class="n">getRandomSplit</span><span class="p">()</span>
</pre></div>
</div>
<p>This loads the Kahn <em>et al.</em> (2014) Polycomb target regions, resizes them to 3kb each, and extracts the underlying sequences from the fruit fly genome. Information about the loaded and processed data can be output by running the following.</p>
<p>Run (optional):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">Kahn2014Rgn</span><span class="o">.</span><span class="n">printStatistics</span><span class="p">()</span>
<span class="n">tpos</span><span class="o">.</span><span class="n">printStatistics</span><span class="p">()</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Kahn2014Rgn</span><span class="o">.</span><span class="n">printStatistics</span><span class="p">()</span>
<span class="go">Region set statistics - GFF file: Kahn2014.GFF</span>
<span class="go"> - Regions: 201</span>
<span class="go"> - Region sequences: 2L, 2R, 3L, 3R, 4, X</span>
<span class="go"> - Mean length: 1000.00 nt (min.: 1000 nt - max.: 1000 nt)</span>
<span class="go"> - Total length: 201000 nt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tpos</span><span class="o">.</span><span class="n">printStatistics</span><span class="p">()</span>
<span class="go">Sequence set statistics - Regions set: GFF file: Kahn2014.GFF (randomly recentered - 3000 bp) - from sequence stream: FASTA sequence stream&lt;dmel-all-chromosome-r5.57.fasta&gt; (split A)</span>
<span class="go"> - Sequences: 100</span>
<span class="go"> - Mean length: 3000.00 nt (min.: 3000 nt - max.: 3000 nt)</span>
<span class="go"> - Total length: 300000 nt</span>
</pre></div>
</div>
<p>We will now generate negative training sequences. To this end, we train a 4th order Markov chain on the PREs, as we did previously (Bredesen <em>et al.</em> 2019). Sequences generated by a 4th order Markov chain trained on PREs retains motif-enrichment, but lack the pairing that is characteristic of PREs.</p>
<p>Run:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">MCPRE</span> <span class="o">=</span> <span class="n">cis</span><span class="o">.</span><span class="n">generatorMarkovChain</span><span class="p">(</span><span class="n">trainingSequences</span> <span class="o">=</span> <span class="n">Kahn2014Seq</span><span class="p">,</span> <span class="n">degree</span> <span class="o">=</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">tneg</span> <span class="o">=</span> <span class="n">MCPRE</span><span class="o">.</span><span class="n">generateSet</span><span class="p">(</span><span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">tpos</span><span class="p">),</span> <span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">tpos</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<span class="n">vneg</span> <span class="o">=</span> <span class="n">MCPRE</span><span class="o">.</span><span class="n">generateSet</span><span class="p">(</span><span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">vpos</span><span class="p">)</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span> <span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">vpos</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
</pre></div>
</div>
<p>This trains the Markov chain a Markov chain on the Kahn <em>et al.</em> (2014) set, and randomly generates a set of as many sequences as are in the Kahn <em>et al.</em> (2014) set, with the same length as the first sequence.</p>
<p>Run (optional):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tneg</span><span class="o">.</span><span class="n">printStatistics</span><span class="p">()</span>
<span class="n">tneg</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
</pre></div>
</div>
<p>Output (will vary depending on random seed):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tneg</span><span class="o">.</span><span class="n">printStatistics</span><span class="p">()</span>
<span class="go">Sequence set statistics - Generated set &lt;Sequences: 100; Length each: 3000; Model: Markov Chain&lt;Degree: 4; Pseudocounts: 1; Add reverse complements: Yes; Training set: Sequence list&lt;Regions set: GFF file: Kahn2014.GFF (randomly recentered - 3000 bp) - from sequence stream: FASTA sequence stream&lt;dmel-all-chromosome-r5.57.fasta&gt;&gt;&gt;; Seed: 182579965&gt;</span>
<span class="go"> - Sequences: 100</span>
<span class="go"> - Mean length: 3000.00 nt (min.: 3000 nt - max.: 3000 nt)</span>
<span class="go"> - Total length: 300000 nt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tneg</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="go">&#39;GTCCCTCGCTCTTTTTACTTTTTAATTGATGATCGCTTATTAAAACTACGCGTTGCCGTTACGAATTGATGTAACTCAGAGATCGGCGCACAAATAATGCCCCTTGGTGGCGTCGAGCACCATCAACGCGAGCGACATAACCCTTTAACTGCGAAAAAAAGGCAAAAGGCAGCTTGTGAGAGCCACTTGCATAGTGCGGAGCACTGCATTGTATGCGGGGAACTCAATGCTAAAGAATAATTAAGTTCTACCGCATGAATATGCCGGGAATAGATTGTATATCTGGTCCCAAGTTCATAACTGCACATTTGTTGTCGAAATAATGAAGGGATCACCAAACTTAGCTCGGTTATATTACACTCTCTGTTTGGCGAGCGGCCACCGATGTCCCTCCCCAAGCCAGGTTCCTGCAATTTGCCTTCCGCACAGATAACGTTATATACGGACCAAGCCACGCTTACTTTAGTTGTCTAACAGACATGGAAAACAAACTTTTGTGACACTTCCAATTAATTTTGGAGGGAAATTTTCGAGCTCGCTTTGATCTACTACTCTCATTAATGAAGTGAGGACAACAGGCACCTTGATGATACATGTATGCAGCGAAGTTACTTGCCCGGTAATTTGATTAGATTCAGTGAAGTTTGGTGGCTGGAGCAGTAATTTGTTGTTGGCATGTTTATAGGGGCCTCCCAAAATTACTCATGAATGGGTCAGGGCATTGACTAGTTAAATTCTTAGCGACAAAAATCATTCCACGCGCATGTATTGGTTTCCTTACAACTGTGCACTGCGGGAGGTAGCCGCATTTTCACCATTTTACAACAGTGGTTCAGGTCGACGGTTGCTCTCCTGAGTAGATATTTCGCTCCCTCTGGTCCAACTTATCGAACCGAAAAATCGACATTATTAAAACAAAACAGCATAAGAAACACTTTGAGTGGGTAAAATGGGAAATTTTTCGTAAATCAACAACCGCCCCAAAGTTCGGCCAATTCGAGCGAGTCGTAGGCCACATTTGTTCGAAGCACACGCATCGCAGCGGTGTGGCTGCAAAACCATCTCATTCGAAGAACTAACGCTCTTCCTTAATAAGTGTAATAAAATACGCGCAAAAATGGCTGCACTGATTTACTTTTCGAGATTTCAAATGAGCCGTAAACGAGTGGTGGTGCGCCAATTGATGGAGCGAGTGATTAACTGTTATAGCAATAAAAAGCGGCAACCCCTTGTTGACGGTTTTTGTCGGAATCTCCATAGCAGTGGGTGTGGTGCCTGTCGCTGACTTTTTCTGACAAAGACCACAGCATTCGTCTTATGGTGTGCGCAATCCCGCTTGTTCATTTGAAACCGGCTTCTTTTTAATGCCGAAAGGGTAATCATCCAAGCACAATTATTGTTGCGGGGCTAAGTCTCTCTTCTCGTATAAAACAATACGGTGGGGTCATGAGACATATAGTTGTGAGCCCGTGGCTTGTCAGACCTCCCTCTGTGGCTGCGTATCAAATTTCTGTCGGGGGTCGGACAATTCAGAATATCCCGGGAACAGATACGCTGTGGTGGCACGCCTGCCGGCAACCTTCCGGTTATGTGCAGCCGTTTCAGGAGACTTGTAGAAGTGCGCGTTCTATCAGTTTGGGCAACGAACCCAGTGCGATTGGCGGTAAAGCGTTTTTTTCACATTTGTTAATAATATTGTTTATTATTTCTTGTCCAGCTGGATAACTTAAATACACGTCGCAGTTGGCAGTGTTTCCATTATTAGTTTTGCTTTCTGAGAAAGTTATTTTGTGATGCAGAAAGAGCACTTCGTCTGTGCCTGTTCCAAGCTATCTCCCCGAACTACTTAGTGTGTTGCACCGTAATACCGCCCGAGTTGTCCACCGATTTGCGAGATACATGGGAGGAATTAAAACACTTTAATCACAAAAAGTTCTTCAGATTTAAATTAGACACACCGAAAGAAATTCATAAAATGCCGGCCAAATCACCTTCTGTGCAAAAATAATGGTTATTCGAGCGGACTCTTAATATTTCAAGCGAGAGCCGCCAGCAGATGCTCGTCTAAAGCCCCCCCTTGTTCATTTGCGGTTTTGATGGGATTTATAGTCACACTGAATATAATCCAGCGGCGTTGTCCTCTGCCTTCGTCTTTGGTGCACAAGCTTGAAACAGTCAAGAGATAAACTATAAACCCTCGGGTGGCCGTCGCCTCGGGGGCACTGTGGGTGTGTGTGGACAAATGGCACACGCAAATTTCCGATCAACATCTGTCGGAGCATTTAAACTCTTAGCGCGTCCCTTGCAAGGAGGAACCATTTTTCATGAAACATAGCTACATATTATGAGGCACTGTGAGGGTGGGGGAAAGTATGACAGATTAATCTTTACATCAATTTGCAGTGAAAGCAAATATTAAATGGCTCCGCCGAATAAATTCGACGAAAAGTGGAGTTGTTACATGTATCCAGTATAAATGCCGCCCGAGAGATTACAATCTTAATCTATCGTTTTGTTCAAATACAAAAAGTGATCCGTCAAATAAACGGTCGTGTAAAACACAGATTCTCCAAGCTGTCCAGAACCACCGCCATTTATATGAGCGACGAAGCACGAAAACGCCGAGTCGCTAGTTCCCGTTTATCTCAAAGCCCTCTCCATTAAAATTTATTTAAATGCACTTTGAAAAAATTAACCAGATTTTTTGATTCCCAGGGTCAGCAACTCACCTGGAGGGGAAAGCGAGTTCTCCCTCTCCAGCATAATAGGTAATAAAAGATCCAACTAACTGGCGTCGCTTCTTTTTTTGTTTTAAATCCTATATTTTTAAATGCAGATGCTTCGTCAATGCTCAAAAATGCGCTGAATATATGCCCCATTAAATACTTGACCATCGAAAATTGTGCCGCCAGGCGTGCACAGCTGCATATGAAAACCTCCCTCCACTAGAAATGCGGATGGTGTCAGCTGCTTTCGCCATTGGGCGAGTTGTTTTGCTCG&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="training-a-basic-model-of-cis-regulatory-module-crm-sequences">
<h2>Training a basic model of <em>cis</em>-regulatory module (CRM) sequences<a class="headerlink" href="#training-a-basic-model-of-cis-regulatory-module-crm-sequences" title="Permalink to this headline">¶</a></h2>
<p>Run:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mdl</span> <span class="o">=</span> <span class="n">cis</span><span class="o">.</span><span class="n">trainPREdictorModel</span><span class="p">(</span><span class="n">motifs</span> <span class="o">=</span> <span class="n">cis</span><span class="o">.</span><span class="n">motifs</span><span class="o">.</span><span class="n">getRingrose2003Motifs_GTGT</span><span class="p">(),</span>
                              <span class="n">positives</span> <span class="o">=</span> <span class="n">tpos</span><span class="p">,</span>
                              <span class="n">negatives</span> <span class="o">=</span> <span class="n">tneg</span><span class="p">)</span>
</pre></div>
</div>
<p>This trains a PREdictor model, using the positives and negatives, and the motifs used in Bredesen <em>et al.</em> (2019). To manually specify motifs, the following syntax is used:</p>
<p>Run:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mdlC</span> <span class="o">=</span> <span class="n">cis</span><span class="o">.</span><span class="n">trainPREdictorModel</span><span class="p">(</span><span class="n">motifs</span> <span class="o">=</span> <span class="n">cis</span><span class="o">.</span><span class="n">motifs</span><span class="p">(</span><span class="s1">&#39;Custom motif set&#39;</span><span class="p">,</span> <span class="p">[</span>
                                  <span class="n">cis</span><span class="o">.</span><span class="n">IUPACMotif</span><span class="p">(</span><span class="s1">&#39;GAF&#39;</span><span class="p">,</span> <span class="s1">&#39;GAGAG&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                                  <span class="n">cis</span><span class="o">.</span><span class="n">IUPACMotif</span><span class="p">(</span><span class="s1">&#39;Z&#39;</span><span class="p">,</span> <span class="s1">&#39;YGAGYG&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                                  <span class="n">cis</span><span class="o">.</span><span class="n">IUPACMotif</span><span class="p">(</span><span class="s1">&#39;Grh&#39;</span><span class="p">,</span> <span class="s1">&#39;TGTTTTTT&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                                  <span class="n">cis</span><span class="o">.</span><span class="n">IUPACMotif</span><span class="p">(</span><span class="s1">&#39;Dsp1&#39;</span><span class="p">,</span> <span class="s1">&#39;GAAAA&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                               <span class="p">]),</span>
                               <span class="n">positives</span> <span class="o">=</span> <span class="n">tpos</span><span class="p">,</span>
                               <span class="n">negatives</span> <span class="o">=</span> <span class="n">tneg</span><span class="p">)</span>
</pre></div>
</div>
<p>The model weights can be inspected.</p>
<p>Run (optional):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mdl</span><span class="o">.</span><span class="n">weights</span>
<span class="n">mdlC</span><span class="o">.</span><span class="n">weights</span>
</pre></div>
</div>
<p>Output (will vary depending on random seed):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mdl</span><span class="o">.</span><span class="n">weights</span>
<span class="go">{Feature&lt;PREdictor motif pair occurrence frequency: En, En (within 219 nt)&gt;: 1.252762968495368, Feature&lt;PREdictor motif pair occurrence frequency: En, G10 (within 219 nt)&gt;: 4.356708826689593, Feature&lt;PREdictor motif pair occurrence frequency: En, GAF (within 219 nt)&gt;: 1.1400690023464555, Feature&lt;PREdictor motif pair occurrence frequency: En, PF (within 219 nt)&gt;: 1.0033021088637848, Feature&lt;PREdictor motif pair occurrence frequency: En, PM (within 219 nt)&gt;: 0.6539264674066638, Feature&lt;PREdictor motif pair occurrence frequency: En, PS (within 219 nt)&gt;: 0.8724881092157624, Feature&lt;PREdictor motif pair occurrence frequency: En, Z (within 219 nt)&gt;: 0.7018808605286999, Feature&lt;PREdictor motif pair occurrence frequency: En, GTGT (within 219 nt)&gt;: 0.7276791641023213, Feature&lt;PREdictor motif pair occurrence frequency: G10, G10 (within 219 nt)&gt;: 4.915591745409362, Feature&lt;PREdictor motif pair occurrence frequency: G10, GAF (within 219 nt)&gt;: 3.029137518883944, Feature&lt;PREdictor motif pair occurrence frequency: G10, PF (within 219 nt)&gt;: 1.7466390339475861, Feature&lt;PREdictor motif pair occurrence frequency: G10, PM (within 219 nt)&gt;: 1.4628344382422251, Feature&lt;PREdictor motif pair occurrence frequency: G10, PS (within 219 nt)&gt;: 1.4621703960985815, Feature&lt;PREdictor motif pair occurrence frequency: G10, Z (within 219 nt)&gt;: 2.204367464848599, Feature&lt;PREdictor motif pair occurrence frequency: G10, GTGT (within 219 nt)&gt;: 2.0703148064157855, Feature&lt;PREdictor motif pair occurrence frequency: GAF, GAF (within 219 nt)&gt;: 1.348646995241774, Feature&lt;PREdictor motif pair occurrence frequency: GAF, PF (within 219 nt)&gt;: 0.5968599611077945, Feature&lt;PREdictor motif pair occurrence frequency: GAF, PM (within 219 nt)&gt;: 0.33332757994174145, Feature&lt;PREdictor motif pair occurrence frequency: GAF, PS (within 219 nt)&gt;: 0.2607262624632529, Feature&lt;PREdictor motif pair occurrence frequency: GAF, Z (within 219 nt)&gt;: 0.8400330761916575, Feature&lt;PREdictor motif pair occurrence frequency: GAF, GTGT (within 219 nt)&gt;: 0.5943726095348998, Feature&lt;PREdictor motif pair occurrence frequency: PF, PF (within 219 nt)&gt;: 0.2363887780642302, Feature&lt;PREdictor motif pair occurrence frequency: PF, PM (within 219 nt)&gt;: 0.3600027340314069, Feature&lt;PREdictor motif pair occurrence frequency: PF, PS (within 219 nt)&gt;: 0.10870420160088923, Feature&lt;PREdictor motif pair occurrence frequency: PF, Z (within 219 nt)&gt;: 0.24686007793152598, Feature&lt;PREdictor motif pair occurrence frequency: PF, GTGT (within 219 nt)&gt;: 0.17918042789309396, Feature&lt;PREdictor motif pair occurrence frequency: PM, PM (within 219 nt)&gt;: 1.0116009116784799, Feature&lt;PREdictor motif pair occurrence frequency: PM, PS (within 219 nt)&gt;: 0.2856639082955432, Feature&lt;PREdictor motif pair occurrence frequency: PM, Z (within 219 nt)&gt;: 0.31462670769523315, Feature&lt;PREdictor motif pair occurrence frequency: PM, GTGT (within 219 nt)&gt;: 0.2975497193848584, Feature&lt;PREdictor motif pair occurrence frequency: PS, PS (within 219 nt)&gt;: 0.09899512179075032, Feature&lt;PREdictor motif pair occurrence frequency: PS, Z (within 219 nt)&gt;: 0.10957358237712889, Feature&lt;PREdictor motif pair occurrence frequency: PS, GTGT (within 219 nt)&gt;: 0.0707518054712688, Feature&lt;PREdictor motif pair occurrence frequency: Z, Z (within 219 nt)&gt;: 0.6523788787562861, Feature&lt;PREdictor motif pair occurrence frequency: Z, GTGT (within 219 nt)&gt;: 0.3766683836192142, Feature&lt;PREdictor motif pair occurrence frequency: GTGT, GTGT (within 219 nt)&gt;: 0.6200698922307692}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mdlC</span><span class="o">.</span><span class="n">weights</span>
<span class="go">{Feature&lt;PREdictor motif pair occurrence frequency: GAF, GAF (within 219 nt)&gt;: 1.348646995241774, Feature&lt;PREdictor motif pair occurrence frequency: GAF, Z (within 219 nt)&gt;: 0.8400330761916575, Feature&lt;PREdictor motif pair occurrence frequency: GAF, Grh (within 219 nt)&gt;: -0.14660347419187536, Feature&lt;PREdictor motif pair occurrence frequency: GAF, Dsp1 (within 219 nt)&gt;: 0.13968471335245347, Feature&lt;PREdictor motif pair occurrence frequency: Z, Z (within 219 nt)&gt;: 0.6523788787562861, Feature&lt;PREdictor motif pair occurrence frequency: Z, Grh (within 219 nt)&gt;: -0.7401041636477159, Feature&lt;PREdictor motif pair occurrence frequency: Z, Dsp1 (within 219 nt)&gt;: -0.01169009500780227, Feature&lt;PREdictor motif pair occurrence frequency: Grh, Grh (within 219 nt)&gt;: 0.5108256237659905, Feature&lt;PREdictor motif pair occurrence frequency: Grh, Dsp1 (within 219 nt)&gt;: -0.2997062654185827, Feature&lt;PREdictor motif pair occurrence frequency: Dsp1, Dsp1 (within 219 nt)&gt;: 0.2312864087094857}</span>
</pre></div>
</div>
</div>
<div class="section" id="predicting-genome-wide">
<h2>Predicting genome-wide<a class="headerlink" href="#predicting-genome-wide" title="Permalink to this headline">¶</a></h2>
<p>In order to make genome-wide predictions, we need to calibrate the threshold of our model. After calibrating, we can apply our model for genome-wide prediction.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calibrate threshold</span>
<span class="n">mdl</span><span class="o">.</span><span class="n">calibrateGenomewidePrecision</span><span class="p">(</span><span class="n">positives</span> <span class="o">=</span> <span class="n">vpos</span><span class="p">,</span>
                                    <span class="n">genome</span> <span class="o">=</span> <span class="n">Dmel</span><span class="p">,</span>
                                    <span class="n">factor</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
                                    <span class="n">precision</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">,</span>
                                    <span class="n">bgModelOrder</span> <span class="o">=</span> <span class="mi">4</span><span class="p">)</span>

<span class="c1"># Predict candidate PREs genome-wide</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">mdl</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">Dmel</span><span class="p">)</span>

<span class="c1"># Save predictions to GFF-file</span>
<span class="n">predictions</span><span class="o">.</span><span class="n">saveGFF</span><span class="p">(</span><span class="s1">&#39;predictions.GFF&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Calibration can take some time. However, on multi-core systems, parallel processing can be enabled by by following command.</p>
<p>Run (optional):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cis</span><span class="o">.</span><span class="n">setNCores</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<p>Run (optional):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mdl</span><span class="o">.</span><span class="n">threshold</span>
<span class="n">predictions</span><span class="o">.</span><span class="n">printStatistics</span><span class="p">()</span>
</pre></div>
</div>
<p>Output (will vary depending on random seed):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mdl</span><span class="o">.</span><span class="n">threshold</span>
<span class="go">260.07098795572915</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictions</span><span class="o">.</span><span class="n">printStatistics</span><span class="p">()</span>
<span class="go">Region set statistics - Predictions</span>
<span class="go"> - Regions: 4740</span>
<span class="go"> - Region sequences: 2L, 2LHet, 2R, 2RHet, 3L, 3LHet, 3R, 3RHet, 4, U, Uextra, X, XHet, YHet</span>
<span class="go"> - Mean length: 941.23 nt (min.: 500 nt - max.: 19960 nt)</span>
<span class="go"> - Total length: 4461430 nt</span>
</pre></div>
</div>
<p>The entire training, calibration and prediction procedure can be written more compactly, as:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cis</span><span class="o">.</span><span class="n">trainPREdictorModel</span><span class="p">(</span><span class="n">motifs</span> <span class="o">=</span> <span class="n">cis</span><span class="o">.</span><span class="n">motifs</span><span class="o">.</span><span class="n">getRingrose2003Motifs_GTGT</span><span class="p">(),</span>
                              <span class="n">positives</span> <span class="o">=</span> <span class="n">tpos</span><span class="p">,</span>
                              <span class="n">negatives</span> <span class="o">=</span> <span class="n">tneg</span><span class="p">)</span>\
         <span class="o">.</span><span class="n">calibrateGenomewidePrecision</span><span class="p">(</span><span class="n">positives</span> <span class="o">=</span> <span class="n">Kahn2014Seq</span><span class="p">,</span>
                              <span class="n">genome</span> <span class="o">=</span> <span class="n">Dmel</span><span class="p">,</span>
                              <span class="n">factor</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
                              <span class="n">precision</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">,</span>
                              <span class="n">bgModelOrder</span> <span class="o">=</span> <span class="mi">4</span><span class="p">)</span>\
         <span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">Dmel</span><span class="p">)</span>\
         <span class="o">.</span><span class="n">saveGFF</span><span class="p">(</span><span class="s1">&#39;predictions.GFF&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="preparing-a-training-set-based-on-genome-wide-experimental-data">
<h2>Preparing a training set based on genome-wide experimental data<a class="headerlink" href="#preparing-a-training-set-based-on-genome-wide-experimental-data" title="Permalink to this headline">¶</a></h2>
<p>When we trained models in the last sections, we used published sets of Polycomb targets. In this section, we will define Polycomb targets based on data from modENCODE.
We have downloaded genome-wide peaks for three <em>D. melanogaster</em> Polycomb group proteins and one histone modification:</p>
<ul class="simple">
<li><p>Pc - Source: ftp://data.modencode.org/D.melanogaster/Non-TF-Chromatin-binding-factor/ChIP-seq/computed-peaks_gff3/Pc%3ADevelopmental-Stage=Embryos-14-16-hr-OR%23Strain=Oregon-R%3AChIP-seq%3ARep-1%3A%3ADmel_r5.32%3AmodENCODE_3957%3A816.gff3.gz</p></li>
<li><p>Psc - Source: source URL: ftp://data.modencode.org/D.melanogaster/Non-TF-Chromatin-binding-factor/ChIP-seq/computed-peaks_gff3/Psc%3ADevelopmental-Stage=Embryos-14-16-hr-OR%23Strain=Oregon-R%3AChIP-seq%3ARep-1%3A%3ADmel_r5.32%3AmodENCODE_3960%3A1817.gff3.gz</p></li>
<li><p>dRING - Source: ftp://data.modencode.org/D.melanogaster/Non-TF-Chromatin-binding-factor/ChIP-seq/computed-peaks_gff3/dRING%3ADevelopmental-Stage=Embryos-14-16-hr-OR%23Strain=Oregon-R%3AChIP-seq%3ARep-1%3A%3ADmel_r5.32%3AmodENCODE_5071%3A1819.gff3.gz</p></li>
<li><p>H3K27me3 - Source URL: tp://data.modencode.org/D.melanogaster/Histone-Modification/ChIP-seq/computed-peaks_gff3/H3K27me3%3ADevelopmental-Stage=Embryos-14-16-hr-OR%23Strain=Oregon-R%3AChIP-seq%3ARep-1%3A%3ADmel_r5.32%3AmodENCODE_3955%3A1820.gff3.gz</p></li>
</ul>
<p>These are included with the tutorial, for educational purposes. All credit for these sets belongs to Karpen G. <em>et al.</em> and modENCODE.</p>
<p>We will now prepare candidate Polycomb targets based on these sets, using the <code class="docutils literal notranslate"><span class="pre">biomarkers</span></code> class in gnocis.</p>
<p>Run:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define the biomarker set, loading signals for each biomarker</span>
<span class="n">PcG</span> <span class="o">=</span> <span class="n">cis</span><span class="o">.</span><span class="n">biomarkers</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;PcG&#39;</span><span class="p">,</span> <span class="n">regionSets</span> <span class="o">=</span> <span class="p">[</span>
	<span class="n">cis</span><span class="o">.</span><span class="n">loadGFFGZ</span><span class="p">(</span><span class="s1">&#39;Pc.gff3.gz&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">getDeltaResized</span><span class="p">(</span><span class="mi">1000</span><span class="p">)</span><span class="o">.</span><span class="n">getRenamed</span><span class="p">(</span><span class="s1">&#39;Pc&#39;</span><span class="p">),</span>
	<span class="n">cis</span><span class="o">.</span><span class="n">loadGFFGZ</span><span class="p">(</span><span class="s1">&#39;Psc.gff3.gz&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">getDeltaResized</span><span class="p">(</span><span class="mi">1000</span><span class="p">)</span><span class="o">.</span><span class="n">getRenamed</span><span class="p">(</span><span class="s1">&#39;Psc&#39;</span><span class="p">),</span>
	<span class="n">cis</span><span class="o">.</span><span class="n">loadGFFGZ</span><span class="p">(</span><span class="s1">&#39;dRING.gff3.gz&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">getDeltaResized</span><span class="p">(</span><span class="mi">1000</span><span class="p">)</span><span class="o">.</span><span class="n">getRenamed</span><span class="p">(</span><span class="s1">&#39;dRING&#39;</span><span class="p">),</span>
	<span class="n">cis</span><span class="o">.</span><span class="n">loadGFFGZ</span><span class="p">(</span><span class="s1">&#39;H3K27me3.gff3.gz&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">getRenamed</span><span class="p">(</span><span class="s1">&#39;H3K27me3&#39;</span><span class="p">)</span>
<span class="p">])</span>

<span class="c1"># Define Highly BioMarker-Enriched loci (HBMEs) based on biomarkers</span>
<span class="n">PcGTargets</span> <span class="o">=</span> <span class="n">PcG</span><span class="o">.</span><span class="n">getHBMEs</span><span class="p">(</span><span class="n">Dmel</span><span class="o">.</span><span class="n">getWindowRegions</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">step</span> <span class="o">=</span> <span class="mi">500</span><span class="p">),</span> <span class="n">threshold</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<p>We here call <code class="docutils literal notranslate"><span class="pre">.getDeltaResized(1000)</span></code> for each DNA-binding factor region set, to account for potential distancing of recruiting sequences and recruited factors. We rename the factors here mostly for cosmetic reasons. When there are multiple files per biomarker to load, renaming them to the same name will merge the signals. This can be useful if incorporating experimental data from several cell types or labs.</p>
<p>Run (optional):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">PcG</span>
<span class="n">PcGTargets</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">PcG</span>
<span class="go">Biomarker set&lt;PcG (H3K27me3 (1486 regions - 1 sets); Pc (1405 regions - 1 sets); Psc (2628 regions - 1 sets); dRING (3473 regions - 1 sets))&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">PcGTargets</span><span class="o">.</span><span class="n">printStatistics</span><span class="p">()</span>
<span class="go">Region set statistics - Windows (Window size: 1000 bp; step size: 500 bp) (PcG HBME)</span>
<span class="go"> - Regions: 1534</span>
<span class="go"> - Region sequences: 2L, 2R, 3L, 3LHet, 3R, 4, U, X</span>
<span class="go"> - Mean length: 7967.41 nt (min.: 1000 nt - max.: 227500 nt)</span>
<span class="go"> - Total length: 12222000 nt</span>
</pre></div>
</div>
<p>We can check the agreement between this set and the Kahn set.</p>
<p>Run (optional):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">PcGTargets</span><span class="o">.</span><span class="n">overlapSensitivity</span><span class="p">(</span><span class="n">Kahn2014Rgn</span><span class="p">)</span>
<span class="n">PcGTargets</span><span class="o">.</span><span class="n">overlapPrecision</span><span class="p">(</span><span class="n">Kahn2014Rgn</span><span class="p">)</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">PcGTargets</span><span class="o">.</span><span class="n">overlapSensitivity</span><span class="p">(</span><span class="n">Kahn2014Rgn</span><span class="p">)</span>
<span class="go">0.9950248756218906</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">PcGTargets</span><span class="o">.</span><span class="n">overlapPrecision</span><span class="p">(</span><span class="n">Kahn2014Rgn</span><span class="p">)</span>
<span class="go">0.08545197740112995</span>
</pre></div>
</div>
<p>Accordingly, the modENCODE-based set covers most of the Kahn <em>et al.</em> (2014) regions, but have low precision to them.</p>
<p>We can also check how sensitive and precise our predictions are to the modENCODE PcG clusters.</p>
<p>Run (optional):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">predictions</span><span class="o">.</span><span class="n">overlapSensitivity</span><span class="p">(</span><span class="n">PcGTargets</span><span class="p">)</span>
<span class="n">predictions</span><span class="o">.</span><span class="n">overlapPrecision</span><span class="p">(</span><span class="n">PcGTargets</span><span class="p">)</span>
</pre></div>
</div>
<p>Output (will vary depending on random seed):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictions</span><span class="o">.</span><span class="n">overlapSensitivity</span><span class="p">(</span><span class="n">PcGTargets</span><span class="p">)</span>
<span class="go">0.3340395480225989</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictions</span><span class="o">.</span><span class="n">overlapPrecision</span><span class="p">(</span><span class="n">PcGTargets</span><span class="p">)</span>
<span class="go">0.17236286919831223</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Bredesen <em>et al.</em> 2019: https://academic.oup.com/nar/article/47/15/7781/5538007</p></li>
<li><p>Kahn <em>et al.</em> 2014: https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1004495</p></li>
<li><p>FlyBase - : https://academic.oup.com/nar/article/41/D1/D751/1051942</p></li>
<li><p>Ringrose <em>et al.</em> 2003: https://www.sciencedirect.com/science/article/pii/S153458070300337X</p></li>
<li><p>ModENCODE:</p></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/gnocis_icon.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Gnocis tutorial</a><ul>
<li><a class="reference internal" href="#set-up">Set-up</a></li>
<li><a class="reference internal" href="#preparing-training-data-using-gnocis">Preparing training data using Gnocis</a></li>
<li><a class="reference internal" href="#training-a-basic-model-of-cis-regulatory-module-crm-sequences">Training a basic model of <em>cis</em>-regulatory module (CRM) sequences</a></li>
<li><a class="reference internal" href="#predicting-genome-wide">Predicting genome-wide</a></li>
<li><a class="reference internal" href="#preparing-a-training-set-based-on-genome-wide-experimental-data">Preparing a training set based on genome-wide experimental data</a></li>
<li><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="license.html"
                        title="previous chapter">License</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="gnocis.html"
                        title="next chapter">gnocis</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/tutorial.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="gnocis.html" title="gnocis"
             >next</a> |</li>
        <li class="right" >
          <a href="license.html" title="License"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Gnocis 0.9.0b documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Bjørn Bredesen.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.1.
    </div>
  </body>
</html>